lib/session_encryptor.rb

```diff
@@ -1,18 +1,154 @@
+  # personal keys are generated and stored in the session between requests, but are used
+  # to decrypt PII bundles, so we treat them similarly to the PII itself.
+  SENSITIVE_PATHS = [
+    ['warden.user.user.session', 'idv/doc_auth'],
+    ['warden.user.user.session', 'idv'],
+    ['warden.user.user.session', 'personal_key'],
+    ['flash', 'flashes', 'personal_key'],
+  ]
+
   def load(value)
-    decrypted = encryptor.decrypt(value)
+    return LegacySessionEncryptor.new.load(value) if should_use_legacy_encryptor_for_read?(value)
+
+    _v2, ciphertext = value.split(':')
+    decrypted = outer_encryptor.decrypt(ciphertext)
+
+    session = JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
+    kms_decrypt_sensitive_paths!(session)
 
-    JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
+    session
   end
 
   def dump(value)
+    return LegacySessionEncryptor.new.dump(value) if should_use_legacy_encryptor_for_write?
+    value.deep_stringify_keys!
+
+    kms_encrypt_pii!(value)
+    kms_encrypt_sensitive_paths!(value, SENSITIVE_PATHS)
+    alert_or_raise_if_contains_sensitive_keys!(value)
     plain = JSON.generate(value, quirks_mode: true)
-    encryptor.encrypt(plain)
+    NEW_CIPHERTEXT_HEADER + ':' + outer_encryptor.encrypt(plain)
+  end
+
+  def kms_encrypt(text)
+    Base64.encode64(Encryption::KmsClient.new.encrypt(text, 'context' => 'session-encryption'))
+  end
+
+  def kms_decrypt(text)
+    Encryption::KmsClient.new.decrypt(
+      Base64.decode64(text), 'context' => 'session-encryption'
+    )
+  end
+
+  def outer_encryptor
+    Encryption::Encryptors::AttributeEncryptor.new
   end
 
   private
 
-  def encryptor
-    Encryption::Encryptors::SessionEncryptor.new
+  # The PII bundle is stored in the user session in the 'decrypted_pii' key.
+  # The PII is decrypted with the user's password when they successfully submit it and then
+  # stored in the session.  Before saving the session, this method encrypts the PII with KMS and
+  # stores it in the 'encrypted_pii' key.
+  #
+  # The PII is not frequently needed in its KMS-decrypted state. To reduce the
+  # risks around holding plaintext PII in memory during requests, this PII is KMS-decrypted
+  # on-demand by the Pii::Cacher.
+  def kms_encrypt_pii!(session)
+    return unless session.dig('warden.user.user.session', 'decrypted_pii')
+    decrypted_pii = session['warden.user.user.session'].delete('decrypted_pii')
+    session['warden.user.user.session']['encrypted_pii'] =
+      kms_encrypt(decrypted_pii)
+    nil
+  end
+
+  # This method extracts all of the sensitive paths that exist into a
+  # separate hash.  This separate hash is then encrypted and placed in the session.
+  # We use #reduce to build the nested empty hash if needed. If Hash#bury
+  # (https://bugs.ruby-lang.org/issues/11747) existed, we could use that instead.
+  def kms_encrypt_sensitive_paths!(session, sensitive_paths)
+    sensitive_data = {
+    }
+
```

Comment:
supernit, I'd initialize this on one line if we don't have any keys to put in it
```suggestion
    sensitive_data = {}

```